### PR TITLE
Add split join integration test

### PR DESCRIPTION
**Pull request by nolash**
An integration test making the necessary adjustments for the joiner to join what the splitter splits.

### CODE REVIEW

File: .githooks/pre-push.bash [hunk 1]
```diff
@@ -1,15 +0,0 @@
-#!/usr/bin/env bash
```
Comment (acud) [inline .githooks/pre-push.bash:1]: @nolash why is this removed? can you please rebase master?

Comment (nolash) [reply .githooks/pre-push.bash:1]: @acud currently this hook fails on my system, and aborts the push.

File: pkg/file/file_test.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+	if l != int64(len(data)) {
+		t.Fatalf("data length return expected %d, got %d", len(data), l)
+	}
+
+	// read from joiner
+	var resultData []byte
+	for i := 0; i < len(data); i += swarm.ChunkSize {
```
Comment (acud) [inline pkg/file/file_test.go:71]: `ioutil.ReadAll()`?

Comment (nolash) [reply pkg/file/file_test.go:71]: Sorry, no; joiner needs chunk-wise reads.

File: pkg/file/joiner/internal/job.go [hunk 3]
```diff
@@ -110,12 +107,24 @@ func (j *SimpleJoinerJob) nextReference(level int) error {
...
 	err := j.nextChunk(level-1, chunkAddress)
 	if err != nil {
-		return err
+		// if the last write is a "dangling chunk" the data chunk will have been moved
+		// to an intermediate level. In this edge case, the error must be suppressed,
+		// and the cursor manually to data length boundary to terminate the loop in
+		// the calling frame.
+		if j.readCount+int64(len(data)) == j.spanLength {
+			j.cursors[level] = len(j.data[level])
+			//j.dataC <- data
```
Comment (acud) [inline pkg/file/joiner/internal/job.go:116]: rm

File: pkg/file/joiner/internal/job.go [hunk 4]
```diff
@@ -110,12 +107,24 @@ func (j *SimpleJoinerJob) nextReference(level int) error {
...
+			err = j.sendChunkToReader(data)
```
Comment (acud) [inline pkg/file/joiner/internal/job.go:117]: `return j.sendChunk...`?

File: pkg/file/joiner/internal/job.go [hunk 5]
```diff
@@ -110,12 +107,24 @@ func (j *SimpleJoinerJob) nextReference(level int) error {
...
+			return err
+		}
+		if err != io.EOF {
```
Comment (acud) [inline pkg/file/joiner/internal/job.go:120]: don't you want to fail-fast here and return the error directly?

File: pkg/file/joiner/internal/job.go [hunk 6]
```diff
@@ -110,12 +107,24 @@ func (j *SimpleJoinerJob) nextReference(level int) error {
...
+			j.logger.Debugf("error in chunk join for %v: %v", chunkAddress, err)
+		}
 	}
 
 	// move the cursor to the next reference
 	j.cursors[level] += swarm.SectionSize
-	return nil
+	return err
```
Comment (acud) [inline pkg/file/joiner/internal/job.go:127]: there should not be an error if the execution gets here. happy path no?

File: pkg/file/splitter/internal/job.go [hunk 7]
```diff
@@ -139,7 +140,9 @@ func (s *SimpleSplitterJob) sumLevel(lvl int) ([]byte, error) {
 	}
 	ref := s.hasher.Sum(nil)
 	addr := swarm.NewAddress(ref)
-	ch := swarm.NewChunk(addr, s.buffer[s.cursors[lvl+1]:s.cursors[lvl]])
+	spanBytes := make([]byte, 8)
+	binary.LittleEndian.PutUint64(spanBytes, uint64(span))
+	ch := swarm.NewChunk(addr, append(spanBytes, s.buffer[s.cursors[lvl+1]:s.cursors[lvl]]...))
```
Comment (acud) [inline pkg/file/splitter/internal/job.go:145]: i have vague recollection of a poor blind perl developer uttering something about heads, tails and appends... i'll know to ignore them next time

File: pkg/file/file_test.go (hunk 2 above)
Comment (acud) [reply pkg/file/file_test.go:71]: hmmm. it would be nice if `Read` is then able to support this somehow. did you look at `ReadAll()` and why would it not work?

File: .githooks/pre-push.bash (hunk 1 above)
Comment (vandot) [reply .githooks/pre-push.bash:1]: @nolash you could only remove hook from `.git/hooks/pre-push`

File: pkg/file/splitter/internal/job.go (hunk 7 above)
Comment (nolash) [reply pkg/file/splitter/internal/job.go:145]: touché, mon frére.

File: pkg/file/joiner/internal/job.go (hunk 6 above)
Comment (nolash) [reply pkg/file/joiner/internal/job.go:127]: not quite;

We want the error to reach the end of the go-routine in `NewSimpleJoinerJob`, so that `Read` calls after EOF actually can return an `EOF` error.

File: pkg/file/joiner/internal/job.go (hunk 4 above)
Comment (nolash) [reply pkg/file/joiner/internal/job.go:117]: I would like to log the error if it's not EOF

File: pkg/file/joiner/internal/job.go (hunk 6 above)
Comment (nolash) [reply pkg/file/joiner/internal/job.go:127]: ... which I hereby did.

File: pkg/file/file_test.go (hunk 2 above)
Comment (nolash) [reply pkg/file/file_test.go:71]: Also please consider; if this is a 100GB file, for example, would you want to ReadAll it at all?

Comment (nolash) [reply pkg/file/file_test.go:71]: Not sure if I understand this.

What would the end-user interface to the reads be, in a non-test scenario?

Comment (janos) [reply pkg/file/file_test.go:71]: Actually, with the amount of data in this test, it is not needed to have readers. But, once we have tests that are validating and measuring performance on large files, we may want to use readers approach. It is fine to keep these data slices in memory.

**Comment by nolash** [comment]
spoke with @janos on phone, agreement that fixed length data in memory is ok for this iteration.

File: pkg/file/joiner/internal/job.go (hunk 6 above)
Comment (acud) [reply pkg/file/joiner/internal/job.go]: Not sure why this was marked as resolved. I see now that the logic of this function has changed extensively...

File: pkg/file/joiner/internal/job.go [hunk 8]
```diff
@@ -110,6 +107,19 @@ func (j *SimpleJoinerJob) nextReference(level int) error {
...
+		// if the last write is a "dangling chunk" the data chunk will have been moved
+		// to an intermediate level. In this edge case, the error must be suppressed,
+		// and the cursor manually to data length boundary to terminate the loop in
+		// the calling frame.
+		if j.readCount+int64(len(data)) == j.spanLength {
+			j.cursors[level] = len(j.data[level])
+			err = j.sendChunkToReader(data)
+			return err
+		}
+		j.logger.Debugf("error in chunk join for %v: %v", chunkAddress, err)
```
Comment (janos) [inline pkg/file/joiner/internal/job.go:122]: Error should not be returned and logged. I am suggesting to handle the error, and log it if necessary by the topmost caller in the execution stack. With both return and log, the same error may be logged multiple times.

Comment (nolash) [reply pkg/file/joiner/internal/job.go:122]: Ok I will _wrap_ the error then. The problem was the store conceals the detail; it just says `not found` and doesn't output the address. Also warrants the question why shouldn't it include the address in the error from the store...?

Comment (janos) [reply pkg/file/joiner/internal/job.go:122]: Yes, that may be useful. We will have to go through storage error messages and logging to clean and improve them.

Comment (nolash) [reply pkg/file/joiner/internal/job.go:122]: so @janos should we instead just add  the address to the error from the store?

Comment (janos) [reply pkg/file/joiner/internal/job.go:122]: I think that it is ok to make this small improvement in this PR and have another one which would address the whole storage.

**Comment by acud** [comment]
@nolash on a sidenote can you please also have a look at the integration of splitter/joiner to the http api?
